### PR TITLE
feat(aws-ec2): control over VPC AZs

### LINKED ISSUE #5847: Control over VPC AZs
## :question: General Issue

Attempting to create a VPC in ap-northeast-1 with 2 AZs and one NatGateway. Upon deploy the stack fails with the error "Nat Gateway is not available in this availability zone"

apne1-az3 in this region does not have NATGateway support, unfortunately it is mapped to AZ ap-northeast-1a in my account.

### example:
```python
vpc = aws_ec2.Vpc(
            self,
            id='prod_vpc',
            cidr='10.199.0.0/16',
            enable_dns_hostnames=False,
            enable_dns_support=True,
            nat_gateways=1,
            max_azs=2,
            subnet_configuration=[
                aws_ec2.SubnetConfiguration(
                    cidr_mask=24,
                    name='public',
                    subnet_type=aws_ec2.SubnetType.PUBLIC
                ),
                aws_ec2.SubnetConfiguration(
                    cidr_mask=20,
                    name='application',
                    subnet_type=aws_ec2.SubnetType.PRIVATE
                )
            ]
        )
```

Running this in a different account in the same region can also have the same issue.

## expected behavior:

If a NATGateway is requested then the CDK should check for support in the AZ before selection. It appears that the CDK is processing in logical order?

## question:

Is there a workaround for this? 

## Environment

  - **CDK CLI Version:** 1.21
  - **OS:** OSX Catalina
  - **Language:** Python

### PR DESCRIPTION
**Pull request by philipmw**
With this change, the `Vpc` construct gains a new constructor prop,
`availabilityZones`, which gives more control over AZs than the existing
`maxAzs` prop.

closes #5847

----

### All Submissions:

* [x] Have you followed the guidelines in our [Contributing guide?](https://github.com/aws/aws-cdk/blob/master/CONTRIBUTING.md)

### Adding new Unconventional Dependencies:

* [ ] This PR adds new unconventional dependencies following the process described [here](https://github.com/aws/aws-cdk/blob/master/CONTRIBUTING.md/#adding-new-unconventional-dependencies)

### New Features

* [x] Have you added the new feature to an [integration test](https://github.com/aws/aws-cdk/blob/master/INTEGRATION_TESTS.md)?
	* [x] Did you use `yarn integ` to deploy the infrastructure and generate the snapshot (i.e. `yarn integ` without `--dry-run`)?

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-ec2/test/vpc.test.ts [hunk 1]
```diff
@@ -588,6 +588,30 @@ describe('vpc', () => {
...
+      const stack = getTestStack();
+      new Vpc(stack, 'VPC', {
+        availabilityZones: [`${stack.region}b`],
+      });
+      Template.fromStack(stack).resourceCountIs('AWS::EC2::Subnet', 2);
+      for (let i = 1; i < 2; i++) {
+        Template.fromStack(stack).hasResourceProperties('AWS::EC2::Subnet', {
+          AvailabilityZone: `${stack.region}b`,
+        });
+      }
```
Comment (comcalvi) [inline packages/@aws-cdk/aws-ec2/test/vpc.test.ts:612]: this loop doesn't do anything, should be:

```suggestion
        Template.fromStack(stack).hasResourceProperties('AWS::EC2::Subnet', {
          AvailabilityZone: `${stack.region}b`,
        });
```

File: packages/@aws-cdk/aws-ec2/lib/vpc.ts [hunk 2]
```diff
@@ -1317,10 +1332,13 @@ export class Vpc extends VpcBase {
 
     Tags.of(this).add(NAME_TAG, props.vpcName || this.node.path);
 
-    this.availabilityZones = stack.availabilityZones;
+    if (props.availabilityZones) {
+      this.availabilityZones = props.availabilityZones;
```
Comment (comcalvi) [inline packages/@aws-cdk/aws-ec2/lib/vpc.ts:1336]: I suspect there is validation that should be done on this parameter, to ensure there are no deployment time issues.

**Review by comcalvi — DISMISSED** [review]
As this is a feature, please add an integ test for this feature.

File: packages/@aws-cdk/aws-ec2/lib/vpc.ts [hunk 3]
```diff
@@ -841,10 +841,21 @@ export interface VpcProps {
...
    * @default 3
    */
   readonly maxAzs?: number;
 
+  /**
+   * Availability zones this VPC spans.
+   *
+   * This option overrides `maxAzs`.
+   *
+   * @default (a subset of) AZs of the stack
```
[the remark this answers is not in the document]
Comment (philipmw) [reply packages/@aws-cdk/aws-ec2/lib/vpc.ts]: Changed

File: packages/@aws-cdk/aws-ec2/test/vpc.test.ts (hunk 1 above)
Comment (philipmw) [reply packages/@aws-cdk/aws-ec2/test/vpc.test.ts]: Oops, that was a thoughtless copy-paste. Fixed.

File: packages/@aws-cdk/aws-ec2/lib/vpc.ts (hunk 2 above)
Comment (philipmw) [reply packages/@aws-cdk/aws-ec2/lib/vpc.ts:1342]: Good idea. Done.

**Comment by philipmw** [comment]
Thanks for the review, @comcalvi . I made the changes you suggested.

File: packages/@aws-cdk/aws-ec2/lib/vpc.ts [hunk 4]
```diff
@@ -841,10 +841,21 @@ export interface VpcProps {
...
+   * Specify this option only if you do not specify `availabilityZones`.
+   *
    * @default 3
    */
   readonly maxAzs?: number;
 
+  /**
+   * Availability zones this VPC spans.
+   *
+   * This option overrides `maxAzs`.
```
Comment (comcalvi) [inline packages/@aws-cdk/aws-ec2/lib/vpc.ts]:
```suggestion
    * Specify this option only if you do not specify `availabilityZones`.
```

File: packages/@aws-cdk/aws-ec2/lib/vpc.ts [hunk 5]
```diff
@@ -1289,6 +1300,10 @@ export class Vpc extends VpcBase {
       throw new Error('To use DNS Hostnames, DNS Support must be enabled, however, it was explicitly disabled.');
     }
 
+    if (props.availabilityZones && props.maxAzs) {
+      throw new Error('Vpc supports either specific \'availabilityZones\' or \'maxAzs\' number, but not both.');
```
Comment (comcalvi) [inline packages/@aws-cdk/aws-ec2/lib/vpc.ts]:
```suggestion
      throw new Error('Vpc supports either specific \'availabilityZones\' or \'maxAzs\', but not both.');
```

**Review by comcalvi — DISMISSED** [review]
Thanks for the PR!

**Comment by philipmw** [comment]
@comcalvi Thanks. I don't know if I messed it up, but the branch was out of date so I merged main into it. Now it needs another review from you. Would you please?

**Comment by AlJohri** [comment]
It would be great to alternatively support `availabilityZoneIds` parameter. This is already supported by `CfnSubnet`: https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_ec2.CfnSubnet.html#availabilityzoneid

**Pull request merged**